com.ibm.streamsx.metrics/impl/java/src/com/ibm/streamsx/metrics/MetricsSink.java

```diff
@@ -0,0 +1,362 @@
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.log4j.Logger;
+import org.elasticsearch.action.index.IndexResponse;
+import org.elasticsearch.client.transport.TransportClient;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.transport.InetSocketTransportAddress;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.transport.client.PreBuiltTransportClient;
+
+import com.ibm.streams.operator.AbstractOperator;
+import com.ibm.streams.operator.OperatorContext;
+import com.ibm.streams.operator.StreamingData.Punctuation;
+import com.ibm.streams.operator.StreamingInput;
+import com.ibm.streams.operator.Tuple;
+import com.ibm.streams.operator.model.InputPortSet;
+import com.ibm.streams.operator.model.InputPortSet.WindowMode;
+import com.ibm.streams.operator.model.InputPortSet.WindowPunctuationInputMode;
+import com.ibm.streams.operator.model.InputPorts;
+import com.ibm.streams.operator.model.Libraries;
+import com.ibm.streams.operator.model.PrimitiveOperator;
+
+import org.influxdb.InfluxDB;
+import org.influxdb.InfluxDB.ConsistencyLevel;
+import org.influxdb.InfluxDBFactory;
+import org.influxdb.dto.BatchPoints;
+import org.influxdb.dto.Point;
+
+/**
+ * Class for an operator that consumes tuples. 
+ * The following event methods from the Operator interface can be called:
+ * </p>
+ * <ul>
+ * <li><code>initialize()</code> to perform operator initialization</li>
+ * <li>allPortsReady() notification indicates the operator's ports are ready to process and submit tuples</li> 
+ * <li>process() handles a tuple arriving on an input port 
+ * <li>processPuncuation() handles a punctuation mark arriving on an input port 
+ * <li>shutdown() to shutdown the operator. A shutdown request may occur at any time, 
+ * such as a request to stop a PE or cancel a job. 
+ * Thus the shutdown() may occur while the operator is processing tuples, punctuation marks, 
+ * or even during port ready notification.</li>
+ * </ul>
+ * <p>With the exception of operator initialization, all the other events may occur concurrently with each other, 
+ * which lead to these methods being called concurrently by different threads.</p> 
+ */
+@PrimitiveOperator(
+		name="MetricsSink",
+		namespace="com.ibm.streamsx.metrics",
+		description=MetricsSource.DESC_OPERATOR
+		)
+@InputPorts({
+	@InputPortSet(
+			description="Port that ingests tuples",
+			cardinality=1,
+			optional=false,
+			windowingMode=WindowMode.NonWindowed,
+			windowPunctuationInputMode=WindowPunctuationInputMode.Oblivious
+			),
+	@InputPortSet(
+			description="Optional input ports",
+			optional=true,
+			windowingMode=WindowMode.NonWindowed,
+			windowPunctuationInputMode=WindowPunctuationInputMode.Oblivious
+			)
+})
+@Libraries({
+	"opt/downloaded/*"
+	})
+public class MetricsSink extends AbstractOperator {
```

Comment:
Is MetricsSink a good name? On the one hand, a [MetricsSink](http://www.ibm.com/support/knowledgecenter/SSCRJU_4.2.0/com.ibm.streams.toolkits.doc/spldoc/dita/tk$spl/op$spl.adapter$MetricsSink.html) operator already exists. On the other hand, you write data to one or two databases.

It seems as if sending to InfluxDB is optional, why? And why is ElasticSearch mandatory?

I wonder whether it makes sense to (a) split the operator - ElasticSearchSink for [ElasticSearch](https://www.elastic.co/products/elasticsearch) and InfluxDBSink for [InfluxDB](https://www.influxdata.com/time-series-platform/influxdb/), and (b) to follow a more generic approach, which means: evaluate the input stream's meta data and send the complete tuple to the underlying database.

From what I can see in the code, the key is the attribute name except for the metricName-metricValue pair, and the value is always the attribute value except for lastTimeRetrieved, which is handled as timestamp. Is there a generic approach? For example, introduce a parameter to name the timestamp attribute. For the metricName-metricValue pair I have no idea for a generic approach so far, but perhaps you have one!?

If we want to stay with metrics-specific sinks, perhaps MetricsElasticSearchSink and MetricsInfluxDBSink might be useful names.